File: game/converter.py
import os
import struct

import pygame

# Cache decoded surfaces per file path to avoid re-reading the same file
_surface_cache = {}
_oos_cache = {}


def _get_int(data, offset):
    return struct.unpack_from('<i', data, offset)[0]


def _get_short(data, offset):
    return struct.unpack_from('<h', data, offset)[0]


def _palette_from_data(data, offset):
    """Read 256 RGB colors (3 bytes each) → list of (r, g, b, a) tuples."""
    colors = []
    for _ in range(256):
        colors.append((data[offset], data[offset + 1], data[offset + 2], 255))
        offset += 3
    return colors


def _palette_from_alpha_data(data, offset):
    """Read 256 colors from 4-byte little-endian values (alpha forced to 255)."""
    # Java: getInt(data, offset) | 0xFF000000 stored as ARGB packed int
    # ARGB: A=FF, R=data[offset+2], G=data[offset+1], B=data[offset]
    colors = []
    for _ in range(256):
        colors.append((data[offset + 2], data[offset + 1], data[offset], 255))
        offset += 4
    return colors


def _make_surface(pixels, width, height):
    surf = pygame.image.frombuffer(bytes(pixels), (width, height), 'RGBA')
    return surf.convert_alpha()


def _cgf_parse_line(pixels, data, idx, length, x, y, palette, width):
    """Parse one CGF scanline. Modifies pixels bytearray in place.

    The Java implementation uses a for(i=0; i<len; i++) loop with an extra i++
    right after reading the opcode. This while loop replicates that behaviour:
    the final `i += 1` at the bottom mirrors the for-loop's own increment.
    """
    i = 0
    while i < length:
        v = data[idx + i]
        i += 1  # skip past opcode (inner i++ from Java)

        if v == 0:
            off = data[idx + i]
            if off == 0:
                return
            x += off
        elif v == 1:  # UNUSED in original game
            count = data[idx + i]
            for _ in range(count):
                i += 1
                pal_idx = data[idx + i]
                pos = x + width * y
                pixels[pos * 4:pos * 4 + 4] = palette[pal_idx]
                x += 1
                i += 1  # extra i++ present in case v=1
        elif v == 2:  # UNUSED in original game
            count = data[idx + i]
            i += 1
            value = data[idx + i]
            i += 1
            # value_alpha = data[idx + i]  # read to advance i, not used
            for _ in range(count):
                pos = x + width * y
                pixels[pos * 4:pos * 4 + 4] = palette[value]
                x += 1
        elif v == 3:
            count = data[idx + i]
            for _ in range(count):
                i += 1
                pal_idx = data[idx + i]
                pos = x + width * y
                pixels[pos * 4:pos * 4 + 4] = palette[pal_idx]
                x += 1
        elif v == 4:
            count = data[idx + i]
            i += 1
            value = data[idx + i]
            for _ in range(count):
                pos = x + width * y
                pixels[pos * 4:pos * 4 + 4] = palette[value]
                x += 1
        else:
            print(f"Unknown byte in CGF: {v} @ {idx + i}")

        i += 1  # outer for-loop increment


def _decode_cgf(file_content, file_path):
    """Decode a CGF sprite file → list of pygame.Surface (one per frame)."""
    parent_dir = os.path.dirname(file_path)
    palette = None

    # Try to find a sibling .til file to use as the palette source
    try:
        for fname in os.listdir(parent_dir):
            if fname.lower().endswith('.til'):
                with open(os.path.join(parent_dir, fname), 'rb') as f:
                    til_data = f.read()
                palette = _palette_from_data(til_data, 0x20)
                break
    except OSError:
        pass

    if palette is None:
        if len(file_content) < 0x400:
            print(f"Cannot find feasible palette for {file_path}")
            return []
        palette = _palette_from_alpha_data(file_content, len(file_content) - 0x400)

    total_frames = _get_int(file_content, 8)
    surfaces = []

    for frame_idx in range(total_frames):
        meta_offset = 7 * 4 + frame_idx * 6 * 4
        width = _get_int(file_content, meta_offset + 8)
        height = _get_int(file_content, meta_offset + 12)
        payload_offset = _get_int(file_content, meta_offset + 20)

        if width < 0 or height < 0:
            print(f"Invalid frame size: {width}x{height}")
            return []

        if width == 0 or height == 0:
            surfaces.append(pygame.Surface((1, 1), pygame.SRCALPHA))
            continue

        pixels = bytearray(width * height * 4)  # RGBA, zero = fully transparent
        start_offset = 7 * 4 + total_frames * 6 * 4 + payload_offset

        for y in range(height):
            line_len = _get_int(file_content, start_offset)
            _cgf_parse_line(pixels, file_content, start_offset + 4, line_len - 4, 0, y, palette, width)
            start_offset += line_len

        surfaces.append(_make_surface(pixels, width, height))

    return surfaces


def _decode_til(file_content, file_path):
    """Decode a TIL tileset file → list of pygame.Surface (one per frame)."""
    palette = _palette_from_data(file_content, 0x20)

    total_width = _get_short(file_content, 8)
    total_height = _get_short(file_content, 10)
    total_width_tiles = file_content[0x15]
    total_height_tiles = file_content[0x17]
    tile_width = total_width // total_width_tiles
    tile_height = total_height // total_height_tiles
    total_frames = _get_short(file_content, 6) + 2
    delta = total_width_tiles * total_height_tiles * 2
    tile_data_offset = total_frames * delta + 0x320

    surfaces = []
    for frame_idx in range(total_frames):
        frame_offset = 0x320 + frame_idx * delta
        pixels = bytearray(total_width * total_height * 4)
        valid = True

        for l in range(total_width_tiles * total_height_tiles):
            tile_id = _get_short(file_content, frame_offset)
            frame_offset += 2
            x_off = (l % total_width_tiles) * tile_width
            y_off = (l // total_width_tiles) * tile_height
            curr_offset = tile_data_offset + tile_id * tile_height * tile_width

            for y in range(tile_height):
                for x in range(tile_width):
                    if curr_offset >= len(file_content):
                        valid = False
                        break
                    color_idx = file_content[curr_offset]
                    curr_offset += 1
                    pos = (x + x_off) + (y + y_off) * total_width
                    pixels[pos * 4:pos * 4 + 4] = palette[color_idx]
                if not valid:
                    break
            if not valid:
                break

        if not valid:
            break

        surfaces.append(_make_surface(pixels, total_width, total_height))

    return surfaces


def _decode_lzp(file_content, file_path):
    """Decode an LZP compressed image file → list of pygame.Surface (one per frame)."""
    if len(file_content) == 0:
        return []

    palette = _palette_from_data(file_content, 0x20)
    total_frames = _get_int(file_content, 0)
    width = _get_int(file_content, 4)
    height = _get_int(file_content, 8)
    N = 4095

    surfaces = []
    for frame_idx in range(total_frames):
        offset_to_offsets = len(file_content) - total_frames * 4 + frame_idx * 4
        offset_local = _get_int(file_content, offset_to_offsets)
        compressed_len = _get_int(file_content, offset_local)

        pixels = bytearray(width * height * 4)  # RGBA, zero = fully transparent
        data_ptr = offset_local + 4
        data_end = data_ptr + compressed_len

        window = [0] * 4096
        dst_pos = 0
        window_index = 4078
        bit_idx = 0
        end = False
        flags = 0

        while not end:
            if bit_idx == 0:
                if data_ptr >= data_end:
                    break
                flags = file_content[data_ptr]
                data_ptr += 1

            if (flags & 1) != 0:
                if data_ptr >= data_end:
                    break
                c = file_content[data_ptr]
                data_ptr += 1
                if dst_pos > width * height:
                    break
                pos = dst_pos % width + (dst_pos // width) * width
                pixels[pos * 4:pos * 4 + 4] = palette[c]
                dst_pos += 1
                window[window_index] = c
                window_index = (window_index + 1) & N
            else:
                if data_ptr >= data_end:
                    break
                offset = file_content[data_ptr]
                data_ptr += 1
                if data_ptr >= data_end:
                    break
                length_byte = file_content[data_ptr]
                data_ptr += 1
                offset |= (length_byte & 0xF0) << 4
                run_len = (length_byte & 0xF) + 3
                for k in range(run_len):
                    c = window[(offset + k) & N]
                    if dst_pos > width * height:
                        end = True
                        break
                    pos = dst_pos % width + (dst_pos // width) * width
                    pixels[pos * 4:pos * 4 + 4] = palette[c]
                    dst_pos += 1
                    window[window_index] = c
                    window_index = (window_index + 1) & N

            flags >>= 1
            bit_idx = (bit_idx + 1) & 7

        surfaces.append(_make_surface(pixels, width, height))

    return surfaces


def decode_oos(file_path):
    """Decode an OOS animation sync file → list of frame indices."""
    if file_path in _oos_cache:
        return _oos_cache[file_path]

    with open(file_path, 'rb') as f:
        data = f.read()

    offset_to_length = _get_int(data, 0x14)
    offset_to_data = _get_int(data, 0x18)
    length = _get_int(data, offset_to_length)
    result = [data[offset_to_data + i] for i in range(length)]

    _oos_cache[file_path] = result
    return result


def decode_surfaces(file_path):
    """Decode a CGF, TIL, or LZP file → list of pygame.Surface.

    Results are cached by file path so repeated calls for the same file
    (e.g. loading different frame ranges from CASELIVE.TIL) are cheap.
    """
    if file_path in _surface_cache:
        return _surface_cache[file_path]

    ext = os.path.splitext(file_path)[1].upper()

    with open(file_path, 'rb') as f:
        file_content = f.read()

    if ext == '.CGF':
        surfaces = _decode_cgf(file_content, file_path)
    elif ext == '.TIL':
        surfaces = _decode_til(file_content, file_path)
    elif ext == '.LZP':
        surfaces = _decode_lzp(file_content, file_path)
    else:
        print(f"Unsupported format for decode_surfaces: {ext}")
        surfaces = []

    _surface_cache[file_path] = surfaces
    return surfaces

File: game-web/src/converter.ts
/**
 * Port of game/converter.py.
 * Decodes CGF, TIL, LZP sprite files and OOS sync files into browser-native types.
 */

// ─── helpers ─────────────────────────────────────────────────────────────────

function getInt(data: Uint8Array, offset: number): number {
  return (
    data[offset] |
    (data[offset + 1] << 8) |
    (data[offset + 2] << 16) |
    ((data[offset + 3] << 24) >> 0)  // sign-extend to match Python struct '<i'
  );
}

function getShort(data: Uint8Array, offset: number): number {
  const v = data[offset] | (data[offset + 1] << 8);
  return v >= 0x8000 ? v - 0x10000 : v;  // sign-extend to match '<h'
}

type RGBA = [number, number, number, number];

function paletteFromData(data: Uint8Array, offset: number): RGBA[] {
  const colors: RGBA[] = [];
  for (let i = 0; i < 256; i++) {
    colors.push([data[offset], data[offset + 1], data[offset + 2], 255]);
    offset += 3;
  }
  return colors;
}

function paletteFromAlphaData(data: Uint8Array, offset: number): RGBA[] {
  const colors: RGBA[] = [];
  for (let i = 0; i < 256; i++) {
    // Java: getInt(data, offset) | 0xFF000000 stored as ARGB packed int
    // ARGB: A=FF, R=data[offset+2], G=data[offset+1], B=data[offset]
    colors.push([data[offset + 2], data[offset + 1], data[offset], 255]);
    offset += 4;
  }
  return colors;
}

function makeBitmap(pixels: Uint8ClampedArray, width: number, height: number): ImageBitmap {
  const imageData = new ImageData(pixels as Uint8ClampedArray<ArrayBuffer>, width, height);
  // createImageBitmap is synchronous-ish via Promise but we pre-await it
  // Caller must await this
  return createImageBitmap(imageData) as unknown as ImageBitmap;
}

async function makeBitmapAsync(pixels: Uint8ClampedArray, width: number, height: number): Promise<ImageBitmap> {
  const imageData = new ImageData(pixels as Uint8ClampedArray<ArrayBuffer>, width, height);
  return createImageBitmap(imageData);
}

// ─── CGF ─────────────────────────────────────────────────────────────────────

function cgfParseLine(
  pixels: Uint8ClampedArray,
  data: Uint8Array,
  idx: number,
  length: number,
  x: number,
  y: number,
  palette: RGBA[],
  width: number,
): void {
  let i = 0;
  while (i < length) {
    const v = data[idx + i];
    i += 1;  // skip past opcode

    if (v === 0) {
      const off = data[idx + i];
      if (off === 0) return;
      x += off;
    } else if (v === 1) {
      const count = data[idx + i];
      for (let c = 0; c < count; c++) {
        i += 1;
        const palIdx = data[idx + i];
        const pos = x + width * y;
        pixels.set(palette[palIdx], pos * 4);
        x += 1;
        i += 1;  // extra i++ present in case v=1
      }
    } else if (v === 2) {
      const count = data[idx + i];
      i += 1;
      const value = data[idx + i];
      i += 1;
      // value_alpha would be data[idx + i]; read to advance i, not used
      for (let c = 0; c < count; c++) {
        const pos = x + width * y;
        pixels.set(palette[value], pos * 4);
        x += 1;
      }
    } else if (v === 3) {
      const count = data[idx + i];
      for (let c = 0; c < count; c++) {
        i += 1;
        const palIdx = data[idx + i];
        const pos = x + width * y;
        pixels.set(palette[palIdx], pos * 4);
        x += 1;
      }
    } else if (v === 4) {
      const count = data[idx + i];
      i += 1;
      const value = data[idx + i];
      for (let c = 0; c < count; c++) {
        const pos = x + width * y;
        pixels.set(palette[value], pos * 4);
        x += 1;
      }
    } else {
      console.warn(`Unknown byte in CGF: ${v} @ ${idx + i}`);
    }

    i += 1;  // outer for-loop increment
  }
}

async function decodeCgf(fileContent: Uint8Array, tilData: Uint8Array | null): Promise<ImageBitmap[]> {
  let palette: RGBA[];

  if (tilData !== null) {
    palette = paletteFromData(tilData, 0x20);
  } else if (fileContent.length >= 0x400) {
    palette = paletteFromAlphaData(fileContent, fileContent.length - 0x400);
  } else {
    console.warn('Cannot find feasible palette for CGF file');
    return [];
  }

  const totalFrames = getInt(fileContent, 8);
  const bitmaps: ImageBitmap[] = [];

  for (let frameIdx = 0; frameIdx < totalFrames; frameIdx++) {
    const metaOffset = 7 * 4 + frameIdx * 6 * 4;
    const width  = getInt(fileContent, metaOffset + 8);
    const height = getInt(fileContent, metaOffset + 12);
    const payloadOffset = getInt(fileContent, metaOffset + 20);

    if (width < 0 || height < 0) {
      console.warn(`Invalid frame size: ${width}x${height}`);
      return [];
    }

    if (width === 0 || height === 0) {
      bitmaps.push(await createImageBitmap(new ImageData(new Uint8ClampedArray(4), 1, 1)));
      continue;
    }

    const pixels = new Uint8ClampedArray(width * height * 4);  // zero = fully transparent
    let startOffset = 7 * 4 + totalFrames * 6 * 4 + payloadOffset;

    for (let y = 0; y < height; y++) {
      const lineLen = getInt(fileContent, startOffset);
      cgfParseLine(pixels, fileContent, startOffset + 4, lineLen - 4, 0, y, palette, width);
      startOffset += lineLen;
    }

    bitmaps.push(await makeBitmapAsync(pixels, width, height));
  }

  return bitmaps;
}

// ─── TIL ─────────────────────────────────────────────────────────────────────

async function decodeTil(fileContent: Uint8Array): Promise<ImageBitmap[]> {
  const palette = paletteFromData(fileContent, 0x20);

  const totalWidth  = getShort(fileContent, 8) & 0xFFFF;
  const totalHeight = getShort(fileContent, 10) & 0xFFFF;
  const totalWidthTiles  = fileContent[0x15];
  const totalHeightTiles = fileContent[0x17];
  const tileWidth  = Math.floor(totalWidth  / totalWidthTiles);
  const tileHeight = Math.floor(totalHeight / totalHeightTiles);
  const totalFrames = (getShort(fileContent, 6) & 0xFFFF) + 2;
  const delta = totalWidthTiles * totalHeightTiles * 2;
  const tileDataOffset = totalFrames * delta + 0x320;

  const bitmaps: ImageBitmap[] = [];

  for (let frameIdx = 0; frameIdx < totalFrames; frameIdx++) {
    let frameOffset = 0x320 + frameIdx * delta;
    const pixels = new Uint8ClampedArray(totalWidth * totalHeight * 4);
    let valid = true;

    for (let l = 0; l < totalWidthTiles * totalHeightTiles; l++) {
      const tileId = getShort(fileContent, frameOffset) & 0xFFFF;
      frameOffset += 2;
      const xOff = (l % totalWidthTiles) * tileWidth;
      const yOff = Math.floor(l / totalWidthTiles) * tileHeight;
      let currOffset = tileDataOffset + tileId * tileHeight * tileWidth;

      outer: for (let y = 0; y < tileHeight; y++) {
        for (let x = 0; x < tileWidth; x++) {
          if (currOffset >= fileContent.length) { valid = false; break outer; }
          const colorIdx = fileContent[currOffset++];
          const pos = (x + xOff) + (y + yOff) * totalWidth;
          pixels.set(palette[colorIdx], pos * 4);
        }
      }
      if (!valid) break;
    }

    if (!valid) break;
    bitmaps.push(await makeBitmapAsync(pixels, totalWidth, totalHeight));
  }

  return bitmaps;
}

// ─── LZP ─────────────────────────────────────────────────────────────────────

async function decodeLzp(fileContent: Uint8Array): Promise<ImageBitmap[]> {
  if (fileContent.length === 0) return [];

  const palette = paletteFromData(fileContent, 0x20);
  const totalFrames = getInt(fileContent, 0);
  const width  = getInt(fileContent, 4);
  const height = getInt(fileContent, 8);
  const N = 4095;

  const bitmaps: ImageBitmap[] = [];

  for (let frameIdx = 0; frameIdx < totalFrames; frameIdx++) {
    const offsetToOffsets = fileContent.length - totalFrames * 4 + frameIdx * 4;
    const offsetLocal = getInt(fileContent, offsetToOffsets);
    const compressedLen = getInt(fileContent, offsetLocal);

    const pixels = new Uint8ClampedArray(width * height * 4);
    let dataPtr = offsetLocal + 4;
    const dataEnd = dataPtr + compressedLen;

    const window = new Uint8Array(4096);
    let dstPos = 0;
    let windowIndex = 4078;
    let bitIdx = 0;
    let end = false;
    let flags = 0;

    while (!end) {
      if (bitIdx === 0) {
        if (dataPtr >= dataEnd) break;
        flags = fileContent[dataPtr++];
      }

      if ((flags & 1) !== 0) {
        if (dataPtr >= dataEnd) break;
        const c = fileContent[dataPtr++];
        if (dstPos > width * height) break;
        const pos = dstPos % width + Math.floor(dstPos / width) * width;
        pixels.set(palette[c], pos * 4);
        dstPos++;
        window[windowIndex] = c;
        windowIndex = (windowIndex + 1) & N;
      } else {
        if (dataPtr >= dataEnd) break;
        const offset = fileContent[dataPtr++];
        if (dataPtr >= dataEnd) break;
        const lengthByte = fileContent[dataPtr++];
        const backOffset = offset | ((lengthByte & 0xF0) << 4);
        const runLen = (lengthByte & 0xF) + 3;
        for (let k = 0; k < runLen; k++) {
          const c = window[(backOffset + k) & N];
          if (dstPos > width * height) { end = true; break; }
          const pos = dstPos % width + Math.floor(dstPos / width) * width;
          pixels.set(palette[c], pos * 4);
          dstPos++;
          window[windowIndex] = c;
          windowIndex = (windowIndex + 1) & N;
        }
      }

      flags >>= 1;
      bitIdx = (bitIdx + 1) & 7;
    }

    bitmaps.push(await makeBitmapAsync(pixels, width, height));
  }

  return bitmaps;
}

// ─── OOS ─────────────────────────────────────────────────────────────────────

export function decodeOos(data: Uint8Array): number[] {
  const offsetToLength = getInt(data, 0x14);
  const offsetToData   = getInt(data, 0x18);
  const length = getInt(data, offsetToLength);
  const result: number[] = [];
  for (let i = 0; i < length; i++) {
    result.push(data[offsetToData + i]);
  }
  return result;
}

// ─── Public API ──────────────────────────────────────────────────────────────

const surfaceCache = new Map<string, ImageBitmap[]>();
const oosCache     = new Map<string, number[]>();

/**
 * Decode a CGF, TIL, or LZP file to ImageBitmap[].
 * Pass the raw file bytes and optionally sibling TIL bytes for palette lookup (CGF only).
 */
export async function decodeSurfaces(
  filePath: string,
  fileContent: Uint8Array,
  tilData: Uint8Array | null = null,
): Promise<ImageBitmap[]> {
  if (surfaceCache.has(filePath)) return surfaceCache.get(filePath)!;

  const ext = filePath.split('.').pop()?.toUpperCase() ?? '';
  let surfaces: ImageBitmap[];

  if (ext === 'CGF') {
    surfaces = await decodeCgf(fileContent, tilData);
  } else if (ext === 'TIL') {
    surfaces = await decodeTil(fileContent);
  } else if (ext === 'LZP') {
    surfaces = await decodeLzp(fileContent);
  } else {
    console.warn(`Unsupported format for decodeSurfaces: ${ext}`);
    surfaces = [];
  }

  surfaceCache.set(filePath, surfaces);
  return surfaces;
}

/** Decode an OOS sync file to a frame-index array. */
export function decodeSyncData(filePath: string, data: Uint8Array): number[] {
  if (oosCache.has(filePath)) return oosCache.get(filePath)!;
  const result = decodeOos(data);
  oosCache.set(filePath, result);
  return result;
}
